[Merged by Bors] - feat(Order): arithmetic of order types

YanYablonovskiy opened the pull request:
Adding the multiplicative and additive monoids on order types, as well as a distributivity law.

---

- [x] depends on: #34034 

<!-- Your PR title will become the first line of the commit message.

In this box, the text above the `---` (if not empty) will be appended
to the commit message, and can be used to give additional context or
details. Please leave a blank newline before the `---`, otherwise GitHub
will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

YanYablonovskiy committed:
Create OrderTypes_basic.lean

YanYablonovskiy committed:
Create OrderTypes_basic.lean

YanYablonovskiy committed:
Create OrderTypes file with definition and some generalisation from Ordinal

YanYablonovskiy committed:
Create OrderTypes file with definition and some generalisation from Ordinal

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy commented:
I tried to get rid of the sorries, they seem so trivial but i couldn't get it

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+-/
+import Mathlib.SetTheory.Cardinal.Basic
+import Mathlib.SetTheory.Ordinal.Basic
+import Mathlib.Order.Category.LinOrd
```
I'm very suspicious of anything outside of category theory (or bespoke algebra beyond my mortal understanding) importing category theory. But maybe there's a good reason; care to explain?

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+import Mathlib.SetTheory.Ordinal.Basic
+import Mathlib.Order.Category.LinOrd
+import Mathlib.Algebra.Order.Ring.Unbundled.Rat
+import Mathlib.Data.Real.Basic
```
This shouldn't import real numbers; we can have a separate file where we define the order type of `ℝ` and prove all of the fancy stuff about η-orders.

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+import Mathlib.SetTheory.Cardinal.Basic
+import Mathlib.SetTheory.Ordinal.Basic
+import Mathlib.Order.Category.LinOrd
+import Mathlib.Algebra.Order.Ring.Unbundled.Rat
```
In fact, maybe this hypothetical η-order file can have all of the material on ℚ as well?

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+/-
+Copyright (c) _
```
Reminder to fill this out; you can copy the header from another file and just modify the necessary fields.

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+/-!
+# OrderTypes
+
+OrderTypes are defined as equivalences of linear orders under order isomorphism. They are endowed
```
```suggestion
Order types are defined as the quotient of linear orders under order isomorphism. They are endowed
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+* `OrderType.mul o₁ o₂`: the product of two OrderTypes `o₁` and `o₂`.
+
+A pre order with a bottom element is registered on OrderTypes, where `⊥` is
+`0`, the OrderType corresponding to the empty type.
```
You can write "order type" as in plaintext, or `OrderType` to clarify you're talking about code.
```suggestion
`0`, the order type corresponding to the empty type.
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+
+/-- Equivalence relation on linear orders on arbitrary types in universe `u`, given by order
+isomorphism. -/
+instance OrderType.instSetoid : Setoid LinOrd where
```
I have a lot to say here. Bear with me.

First of all, I don't think `LinOrd` is the correct tool to use. We don't really care about the categorical properties of linear orders within this file, do we? We'd save on a lot of needless imports if we just make a new wrapper around `LinearOrder`.

But also, I argue we don't even want to do this. `LinearOrder` comes bundled with a lot of structure: a `≤` and `<` relation, `max` and `min` functions, and decidable instances. Truthfully, all we need is a reflexive, transitive, antisymmetric and total relation.

Here's my idea:
```lean 4
/-- An auxiliary structure representing a linearly ordered type. -/
structure LinearOrderedType : Type (u + 1) where
  carrier : Type u
  le : LE carrier
  isLinearOrder_carrier : Std.IsLinearOrder carrier
```
Then, the `OrderType.type` function would have the following signature:
```lean 4
def type (α : Type u) [h₁ : LE α] [h₂ : Std.IsLinearOrder α] : OrderType :=
  ⟦⟨α, h₁, h₂⟩⟧
```
Note that in my proposed design, the relation comes bundled within the `LE α` instance, unlike as in [`Ordinal.type`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/SetTheory/Ordinal/Basic.html#Ordinal.type) which allows for an unbundled relation. This design is deliberate, and I believe the `Ordinal.type` design was a mistake. Realistically if you ever have a relation which behaves like an order relation, let alone a linear order (or a well order!), you should at least have some sort of type alias which makes this relation into a `Preorder`/etc. instance, so as to make the tons of existing machinery on linear orders usable.

Now, for this same reason, working with just a `Std.IsLinearOrder` instance is going to be quite painful. Therefore, I suggest that you keep `OrderType.induction` with the same type signature; that is, you should prove that every order type arises as the order type from a `LinearOrder` instance. This gives you the best of both worlds: you can instantiate an `OrderType` without worrying about all the extra structure that comes with `LinearOrder`, but you then get to use this extra structure when you prove any results about `OrderType`.

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+def type (α : Type u) [LinearOrder α] : OrderType :=
+  ⟦⟨α⟩⟧
+
+instance zero : Zero OrderType where
```
About two years ago, there was some discussion which ultimately culminated in changing the definition of the `Zero Cardinal` instance from `#PEmpty = 0` to #(Fin 0) = 0`. The idea here being that in pen and paper mathematics, one defines the cardinal `n` as the set of ordinals/naturals below `n`.

I don't really agree with this reasoning, and I think it's fine to just define the zero/one cardinals/ordinals using `PEmpty` and `PUnit`, but I figured you should be aware of this.

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+  ⟨0⟩
+
+instance : One OrderType where
+ one := ⟦LinOrd.of PUnit⟧
```
```suggestion
  one := type PUnit
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+    [LinearOrder β] (h : α ≃o β) : type α = type β :=
+  type_eq.2 ⟨h⟩
+
+theorem type_eq_zero_of_empty [LinearOrder α] [IsEmpty α] : type α = 0 := by
```
```suggestion
@[simp]
theorem type_eq_zero [LinearOrder α] [IsEmpty α] : type α = 0 := by
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+ convert (ordIsoOfIsEmpty α PEmpty).ordertype_congr
+
+@[simp]
+theorem type_eq_zero_iff_isEmpty [LinearOrder α] : type α = 0 ↔ IsEmpty α :=
```
```suggestion
theorem type_eq_zero_iff [LinearOrder α] : type α = 0 ↔ IsEmpty α :=
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+  type_eq.2 ⟨h⟩
+
+theorem type_eq_zero_of_empty [LinearOrder α] [IsEmpty α] : type α = 0 := by
+ convert (ordIsoOfIsEmpty α PEmpty).ordertype_congr
```
(Do you even need convert here?)
```suggestion
  convert (ordIsoOfIsEmpty α PEmpty).ordertype_congr
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+    s.toEquiv.isEmpty,
+    @type_eq_zero_of_empty α _⟩
+
+theorem type_ne_zero_iff_nonempty [LinearOrder α] : type α ≠ 0 ↔ Nonempty α := by simp
```
```suggestion
theorem type_ne_zero_iff [LinearOrder α] : type α ≠ 0 ↔ Nonempty α := by simp
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+
+theorem type_ne_zero_iff_nonempty [LinearOrder α] : type α ≠ 0 ↔ Nonempty α := by simp
+
+theorem type_ne_zero_of_nonempty [LinearOrder α] [h : Nonempty α] : type α ≠ 0 :=
```
```suggestion
theorem type_ne_zero [LinearOrder α] [h : Nonempty α] : type α ≠ 0 :=
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+theorem type_empty : type Empty = 0 :=
+  type_eq_zero_of_empty
+
+theorem type_eq_one_of_unique [LinearOrder α] [Nonempty α] [Subsingleton α] : type α = 1 := by
```
```suggestion
@[simp]
theorem type_eq_one [LinearOrder α] [Nonempty α] [Subsingleton α] : type α = 1 := by
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+  exact (@ofUniqueOfIrrefl _).ordertype_congr
+
+@[simp]
+theorem type_eq_one_iff_unique [LinearOrder α] : type α = 1 ↔ Nonempty (Unique α) :=
```
```suggestion
theorem type_eq_one_iff [LinearOrder α] : type α = 1 ↔ Nonempty (Unique α) :=
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+  rfl
+
+@[simp]
+theorem toType_empty_iff_eq_zero {o : OrderType} : IsEmpty o.toType ↔ o = 0 := by
```
```suggestion
theorem isEmpty_toType_iff {o : OrderType} : IsEmpty o.toType ↔ o = 0 := by
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+open Classical in
+/-- To prove a result on OrderTypes, it suffices to prove it for order types of linear orders. -/
+@[elab_as_elim]
+theorem inductionOnLinOrd {C : OrderType → Prop} (o : OrderType)
```
Is this not the exact same as `inductionOn`? Why even have it?

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+  inductionOn o fun α ↦ H α
+
+open Classical in
+/-- To define a function on OrderTypes, it suffices to define them on all linear order isomorphisms.
```
```suggestion
/-- To define a function on `OrderType`, it suffices to define it on all linear orders.
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+  change Quotient.liftOn' ⟦_⟧ _ _ = _
+  rw [Quotient.liftOn'_mk]
+
+/-! ### The order on OrderTypes -/
```
```suggestion
/-! ### The order on `OrderType` -/
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+For `OrderType`:
+
+Less-than-or-equal is defined such that linear orders `r` on `α` and `s` on `β`
+satisfy `type α ≤ type β` if there exists an order embedding from `α` to `β`.
```
I think we can make this a bit less verbose:
```suggestion
The order is defined so that `type α ≤ type β` iff there exists an order embedding `α ↪o β`.
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+    [LinearOrder β] : type α ≤ type β ↔ Nonempty (α ↪o β) :=
+  Iff.rfl
+
+theorem type_le_iff' {α β} [inst1 : LinearOrder α]
```
Isn't this entirely redundant?

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+    [inst2 : LinearOrder β] : type α ≤ type β ↔ Nonempty (inst1.le ↪r inst2.le) :=
+  ⟨fun f ↦ f, fun f ↦ f⟩
+
+section Cardinal
```
I think this should probably go on another file.

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+  rfl
+
+@[gcongr]
+theorem card_le_card {o₁ o₂ : OrderType} : o₁ ≤ o₂ → card o₁ ≤ card o₂ :=
```
Can you also define `card_mono : Monotone card`?

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+theorem card_le_card {o₁ o₂ : OrderType} : o₁ ≤ o₂ → card o₁ ≤ card o₂ :=
+  inductionOn o₁ fun _ ↦ inductionOn o₂ fun _ _ ⟨f⟩ ↦ ⟨f.toEmbedding⟩
+
+theorem ord_card_mono {o₁ o₂ : OrderType} (h : o₁ ≤ o₂) : (card o₁).ord ≤ (card o₂).ord :=
```
This feels unnecessary.

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+
+@[simp]
+lemma type_add (α : Type u) (β : Type v) [LinearOrder α] [LinearOrder β] :
+    type (α ⊕ₗ β) = (type α) + (type β) := rfl
```
```suggestion
    type (α ⊕ₗ β) = type α + type β := rfl
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+   (fun _ _ ha _ _ hb ↦ ⟨OrderIso.sumLexCongr (choice ha) (choice hb)⟩)
+
+open Classical
+in instance : HAdd OrderType.{u} OrderType.{v} OrderType.{max u v} where
```
Pretty interesting choice to make addition heterogeneous. I think this design is likely a good one, but it's also drastically different from the existing design, which is to use lifting functions from one universe to the other. Can you bring this up on Zulip?

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+lemma type_add (α : Type u) (β : Type v) [LinearOrder α] [LinearOrder β] :
+    type (α ⊕ₗ β) = (type α) + (type β) := rfl
+
+lemma OrderIso.sumLexEmpty (α : Type u) [LinearOrder α] : Nonempty (Lex (α ⊕ PEmpty) ≃o α) :=
```
a) Why wrap this in `Nonempty`? You can just construct the explicit isomorphism.
b) Can you generalize `PEmpty` to an `IsEmpty` type?
c) Can you put this in a more appropriate file?
d) Same below.

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+
+open Classical in
+lemma add_assoc (o₁ o₂ o₃ : OrderType.{u}) : o₁ + o₂ + o₃ = o₁ + (o₂ + o₃) :=
+  inductionOn₃ o₁ o₂ o₃ (fun α _ β _ γ _ ↦ RelIso.ordertype_congr (OrderIso.sumLexAssoc α β γ))
```
Can you inline these proofs in the `AddMonoid` instance?

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+  add_assoc := add_assoc
+  zero_add := zero_add
+  add_zero := add_zero
+  nsmul := nsmulRec
```
What about explicitly defining `↑n = type (Fin n)`?

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+
+open Classical in
+instance : HMul OrderType.{u} OrderType.{v} OrderType.{max u v} where
+  hMul := Quotient.map₂ (fun r s ↦ ⟨(r ×ₗ s)⟩)
```
This needs to be the other way around! Unfortunately Cantor really wanted his ordinal functions to all be normal on the right argument, so now we're stuck writing `ω * 2` until the end of time.
```suggestion
  hMul := Quotient.map₂ (fun r s ↦ ⟨(s ×ₗ r)⟩)
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+lemma mul_assoc (o₁ o₂ o₃ : OrderType.{u}) : o₁ * o₂ * o₃ = o₁ * (o₂ * o₃) :=
+  inductionOn₃ o₁ o₂ o₃ (fun α _ β _ γ _ ↦ RelIso.ordertype_congr (OrderIso.prodLexAssoc α β γ))
+
+instance : Monoid OrderType where
```
Again, can you inline the proofs?

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+
+section Ordinal
+
+def LinearOrder.toWellOrder (α : Type u) [LinearOrder α] [IsWellOrder α (· < ·)] : WellOrder :=
```
a) This should go on another file.
b) I don't think we even want this. We should eventually just redefine ordinals as a subtype of `OrderType`.

vihdzp reviewed (commented):
Thanks for working on this! I've got lots of small suggestions/improvements, and a few larger ones.

vihdzp reviewed (commented):
Thanks for working on this! I've got lots of small suggestions/improvements, and a few larger ones.

YaelDillies commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+import Mathlib.Data.Real.Basic
+import Mathlib.Data.Sum.Order
+import Mathlib.Data.Prod.Lex
+import Mathlib.Logic.Equiv.Prod
```
That's definitely quite a few too many imports

YaelDillies commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+instance : Monoid OrderType where
+  mul_assoc := mul_assoc
+  one_mul := one_mul
+  mul_one := mul_one
```
```suggestion
instance : Monoid OrderType where
  mul_assoc o₁ o₂ o₃ := by
    classical
    exact inductionOn₃ o₁ o₂ o₃ (fun α _ β _ γ _ ↦ RelIso.ordertype_congr (OrderIso.prodLexAssoc α β γ))
  one_mul := one_mul
  mul_one := mul_one
```
and same for the other two fields

YaelDillies commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+        ,
+        fun h ↦
+        match a, b, h with
+        | ⟨⟨_ , _⟩ , _⟩ , ⟨⟨_, _⟩ , _⟩ , h1 => by sorry⟩}
```
These should move earlier

YaelDillies commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+def eta : OrderType := type ℚ
+
+/-- The order type of the real numbers on the interval `(0,1)`. -/
+def theta : OrderType := type (Set.Ioo (0 : ℝ) 1)
```
Would be good to double-check these notations are the literature-approved ones

vihdzp replied:
I can vouch for omega and eta, I'd have to check for theta.

YaelDillies replied:
Also surely `type (Set.Ioo (0 : ℝ) 1) = type ℝ`, no?

YaelDillies commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,450 @@
+Copyright (c) _
+-/
+import Mathlib.SetTheory.Cardinal.Basic
+import Mathlib.SetTheory.Ordinal.Basic
```
We already talked about this, but IMO ordinals should import order types, not the other way around

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>
Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>
Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

YanYablonovskiy committed:
Update OrderType.lean

Add copyright
[skip actions]

YanYablonovskiy committed:
Update OrderType.lean

Add copyright
[skip actions]

vihdzp commented:
(once you've applied all/most suggestions, remove the awaiting author tag and please make sure to notify me!)

YanYablonovskiy committed:
Clean up imports, get rid of sorries, and address some comments

YanYablonovskiy committed:
Clean up imports, get rid of sorries, and address some comments

YanYablonovskiy committed:
Bump

YanYablonovskiy committed:
Bump

YanYablonovskiy committed:
Fix module error

YanYablonovskiy committed:
Fix module error

YanYablonovskiy committed:
Style fix

YanYablonovskiy committed:
Style fix

YanYablonovskiy committed:
Apply suggestions from code review

[skip ci]

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>
Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>

YanYablonovskiy committed:
Apply suggestions from code review

[skip ci]

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>
Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>

YanYablonovskiy committed:
Finish conversation review suggestions

YanYablonovskiy committed:
Finish conversation review suggestions

YanYablonovskiy commented:
-awaiting-author

YanYablonovskiy committed:
Minor typos

YanYablonovskiy committed:
Minor typos

YanYablonovskiy committed:
Add doc-string and some more typos

YanYablonovskiy committed:
Add doc-string and some more typos

YanYablonovskiy committed:
Fix recommended_spellings

YanYablonovskiy committed:
Fix recommended_spellings

YaelDillies commented on Mathlib/Order/Hom/Basic.lean:
```diff
@@ -884,6 +884,16 @@ def prodComm : α × β ≃o β × α where
+    map_rel_iff' := fun {a b} ↦ by
+      rcases a with ⟨⟨_ , _⟩ , _⟩
+      rcases b with ⟨⟨_, _⟩ , _⟩
+      simp [Equiv.prodAssoc, and_assoc] }
```
```suggestion
    (α × β) × γ ≃o α × (β × γ) where
  toEquiv := .prodAssoc α β γ
  map_rel_iff' := @fun ⟨⟨_ , _⟩ , _⟩ ⟨⟨_, _⟩ , _⟩ => by simp [Equiv.prodAssoc, and_assoc]
```

YaelDillies commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,397 @@
+instance instNeZeroOne : NeZero (1 : OrderType) :=
+  ⟨OrderType.one_ne_zero⟩
+
+theorem type_le_iff {α β} [LinearOrder α]
```
```suggestion
theorem type_le_type_iff {α β} [LinearOrder α]
```

YaelDillies commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,397 @@
+
+theorem _root_.RelEmbedding.type_le {α β}
+    [LinearOrder α] [LinearOrder β] (h : α ↪o β) : type α ≤ type β :=
+  ⟨h⟩
```
```suggestion
alias _root_.OrderEmbedding.type_le_type  := type_le_type
```

vihdzp replied:
(There's a spurious space in your suggestion)

YaelDillies reviewed (commented):
Could you make a preliminary PR for the order isos?

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,397 @@
+structure LinearOrderedType : Type (u + 1) where
+  carrier : Type u
+  le : LE carrier
+  isLinearOrder_carrier : Std.IsLinearOrder carrier
```
I suggested this design and gave an argument against `LinOrd` [in this reply](https://github.com/leanprover-community/mathlib4/pull/33420?notification_referrer_id=NT_kwDOA-btRrQyMTU3Mzc0OTAwMDo2NTQ2NTY3MA#discussion_r2654872992). In particular:

- I think the categorical properties of `LinOrd` are irrelevant here.
- We don't need the extra structure that `LinearOrder` comes bundled with to define an order type.

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,397 @@
+lemma one_def : (1 : OrderType) = type PUnit := rfl
+
+@[simp]
+theorem type_ordtoType (o : OrderType) : type o.ToType = o := surjInv_eq Quot.exists_rep o
```
I think this name is wrong too, shouldn't it be `type_toType`?

vihdzp commented on Mathlib/Order/Hom/Lex.lean:
```diff
@@ -179,4 +179,28 @@ variable {α β} in
+  { toEquiv := .trans ofLex <| .trans (.prodCongr ofLex <| .refl _) <|
+      .trans (.prodAssoc α β γ) <| .trans (.prodCongr (.refl _) toLex) <| toLex
+    map_rel_iff' := by simp [Prod.Lex.le_iff,Prod.Lex.lt_iff]; grind
+   }
```
```suggestion
    [LinearOrder α] [LinearOrder β] [LinearOrder γ] : (α ×ₗ β) ×ₗ γ ≃o α ×ₗ β ×ₗ γ where
  toEquiv := .trans ofLex <| .trans (.prodCongr ofLex <| .refl _) <|
    .trans (.prodAssoc α β γ) <| .trans (.prodCongr (.refl _) toLex) <| toLex
  map_rel_iff' := by simp [Prod.Lex.le_iff, Prod.Lex.lt_iff]; grind
```
There's a non-terminal `simp` in the last line. Could you fix that? I imagine the fix is to either
- Replace `simp` by `simp only`.
- Pass these predicates to `grind` instead.

vihdzp commented on Mathlib/Order/Hom/Lex.lean:
```diff
@@ -179,4 +179,28 @@ variable {α β} in
+  toEquiv := ofLex.trans ((Equiv.prodCongr ea eb).trans toLex)
+  map_rel_iff' := by
+    intro a b
+    simp [Prod.Lex.le_iff, OrderIso.lt_iff_lt]
```
Does this not work?
```suggestion
  map_rel_iff' := by simp [Prod.Lex.le_iff]
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,397 @@
+* `OrderType`: the type of OrderTypes (in a given universe)
+* `OrderType.type α`: given a type `α` with a linear order, this is the corresponding OrderType,
+* `OrderType.card o`: the cardinality of an OrderType `o`.
+* `o₁ + o₂`: the lexicographic sum of order types, which forms an `AddMonoid`
```
For consistency.
```suggestion
* `o₁ + o₂`: the lexicographic sum of order types, which forms an `AddMonoid`.
```

vihdzp commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,397 @@
+
+/-- `OrderType.{u}` is the type of linear orders in `Type u`, up to order isomorphism. -/
+@[pp_with_univ]
+public def OrderType : Type (u + 1) :=
```
Huh? Shouldn't all of the material in this file be public? Can you wrap this in a `public section`?

vihdzp commented on Mathlib/Order/Hom/Basic.lean:
```diff
@@ -1069,6 +1079,10 @@ noncomputable def ofUnique
 
+/-- The order isomorphism `α ≃o β` when `α` has a linear order and `β` is empty via a predicate. -/
+def ofIsEmpty (α β : Type*) [Preorder α] [IsEmpty β] [LinearOrder β] [IsEmpty α] : α ≃o β :=
+  ⟨Equiv.equivOfIsEmpty α β, @fun a ↦ isEmptyElim a⟩
```
Does this work?
```suggestion
  ⟨Equiv.equivOfIsEmpty α β, isEmptyElim⟩
```

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>
Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>
Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>

YanYablonovskiy commented on Mathlib/Order/OrderType.lean:
```diff
@@ -0,0 +1,397 @@
+
+instance (x : LinearOrderedType) : LE x.carrier := x.le
+
+instance (x : LinearOrderedType) : LinearOrder x.carrier where
```
How about making it `private local` , because this instance is currently the glue that makes all the other proofs need minimal changes

YanYablonovskiy committed:
Apply suggestions

YanYablonovskiy committed:
Apply suggestions

vihdzp commented:
All we actually need from `LinOrd` is the definition. But if we can move it out into some defs file then sure, we might as well use it.

YaelDillies commented:
I think it's reasonable to move it to `Order.Defs`

YanYablonovskiy committed:
Refactor as per suggestions, back to `LinOrd`

YanYablonovskiy committed:
Refactor as per suggestions, back to `LinOrd`

YanYablonovskiy committed:
Change mathlib imports from `OrderType`

`OrderType.lean` no longer exists

YanYablonovskiy committed:
Change mathlib imports from `OrderType`

`OrderType.lean` no longer exists

vihdzp commented on Mathlib/Order/Types/Defs.lean:
```diff
@@ -0,0 +1,229 @@
+  le_trans a b c :=
+    Quotient.inductionOn₃ a b c fun _ _ _ ⟨f⟩ ⟨g⟩ ↦ ⟨f.trans g⟩
+
+instance instNeZeroOne : NeZero (1 : OrderType) :=
```
I'd rather not be naming instances unless we have a legitimate reason to.
```suggestion
instance : NeZero (1 : OrderType) :=
```

vihdzp commented on Mathlib/Order/Types/Defs.lean:
```diff
@@ -0,0 +1,229 @@
+
+alias _root_.OrderEmbedding.type_le_type := type_le_type
+
+protected theorem zero_le (o : OrderType) : 0 ≤ o :=
```
```suggestion
@[simp]
protected theorem zero_le (o : OrderType) : 0 ≤ o :=
```

vihdzp commented on Mathlib/Order/Types/Defs.lean:
```diff
@@ -0,0 +1,229 @@
+-/
+module
+
+public import Mathlib.Order.Hom.Basic
```
Defs files should generally only import other Defs files. What happens if you replace this import by?
```suggestion
public import Mathlib.Order.Hom.Defs
```
Can we still define all the important things in this file? If so, can you move all the remaining theorems to the `Basic` file?

vihdzp commented on Mathlib/Order/Types/Defs.lean:
```diff
@@ -0,0 +1,229 @@
+@[simp]
+protected theorem not_lt_zero (o : OrderType) : ¬o < 0 :=
+  not_lt_bot
+
```
A theorem `pos_iff_ne_zero : 0 < o ↔ o ≠ 0` would be nice too.

YanYablonovskiy replied:
Only managed one direction so far

vihdzp commented on Mathlib/Order/Hom/Basic.lean:
```diff
@@ -884,6 +884,13 @@ def prodComm : α × β ≃o β × α where
+def prodAssoc (α β γ : Type*) [LE α] [LE β] [LE γ] :
+    (α × β) × γ ≃o α × (β × γ) where
+  toEquiv := .prodAssoc α β γ
+  map_rel_iff' := @fun ⟨⟨_ , _⟩ , _⟩ ⟨⟨_, _⟩ , _⟩ => by simp [Equiv.prodAssoc, and_assoc]
```
```suggestion
  map_rel_iff' := @fun ⟨⟨_, _⟩, _⟩ ⟨⟨_, _⟩, _⟩ ↦ by simp [Equiv.prodAssoc, and_assoc]
```

vihdzp commented on Mathlib/Order/Types/Basic.lean:
```diff
@@ -0,0 +1,149 @@
+* `OrderType.card o`: the cardinality of an OrderType `o`.
+* `o₁ + o₂`: the lexicographic sum of order types, which forms an `AddMonoid`.
+* `o₁ * o₂`: the lexicographic product of order types, which forms a `MonoidWithZero`.
+* `OrderType.mul o₁ o₂`: the product of two OrderTypes `o₁` and `o₂`.
```
Is this not redundant?

vihdzp reviewed (commented):
Small nitpicks of mine. This PR is shaping up quite nicely!

You could already start properly PRing the first pieces of this. The material on `Order.Hom` looks quite good (modulo some small suggestions) and I'd like to use it to golf the analogous proofs on `Ordinal`. You could then PR the `Order.Types.Defs` file, and at that point we can start thinking about what kind of API to add where and how to eventually link this to `Ordinal`.

vihdzp commented:
🌶️ take incoming: should we eventually move the `Ordinal` folder to `Order` or even to `Order.Type`? It feels kind of silly to categorize it as "set theory" when our development is not even based on ZFC sets. And it'd be nice to keep these parallel developments (general order types, ordinals) close together.

vihdzp commented on Mathlib/Order/Types/Defs.lean:
```diff
@@ -0,0 +1,229 @@
+/-
```
The `Order` part of `OrderType` seems redundant given that we're already in the `Order` folder.

YaelDillies replied:
But `Order.Type` sounds like it should be about order properties of `Type`

vihdzp replied:
I guess we can also just keep the folder named `Order.Types`. Matches `ModelTheory.Types`, for instance.

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>

YanYablonovskiy commented on Mathlib/Order/Hom/Lex.lean:
```diff
@@ -143,6 +143,16 @@ theorem sumLexIicIoi_symm_apply_Iic (a : Iic x) : (sumLexIicIoi x).symm a = Sum.
 theorem sumLexIicIoi_symm_apply_Ioi (a : Ioi x) : (sumLexIicIoi x).symm a = Sum.inr a :=
   sumLexIicIoi_symm_apply_of_lt a.2
 
+lemma sumLexEmpty (α : Type*) {β : Type*} [LinearOrder α] [IsEmpty β] [LinearOrder β] :
```
are we sure not at least `PartialOrder`

vihdzp replied:
If you want to add this, I think `Mathlib.Data.Sum.Order` might be a more fitting file.

YanYablonovskiy commented on Mathlib/Order/Types/Basic.lean:
```diff
@@ -0,0 +1,149 @@
+
+variable {α : Type u} {β : Type v} {γ : Type w} {δ : Type w'}
+
+instance : ZeroLEOneClass OrderType :=
```
seem to need imports for `ZeroLEOneClass`

YanYablonovskiy committed:
Changes from review

YanYablonovskiy committed:
Changes from review

YanYablonovskiy commented:
Still need to move some of the OrderTypes , and will open a seperate pr for the OrderIso/Lex lemmas.

-awaiting-author

YanYablonovskiy committed:
Code review suggestions

YanYablonovskiy committed:
Code review suggestions

And generalizing results about empty lexicographic sum orders to just less-than (as opposed to `Preorder`).

YanYablonovskiy committed:
Fix name

vihdzp commented:
Could you please resolve the merge conflict?

Also, I believe the material up to the `LeftDistribClass` instance should be Mathlib ready. Could you split that up into a new PR? That way we can review and merge in stages (and I can start using this new type myself!).

vihdzp commented on Mathlib/Order/Types/Defs.lean:
```diff
@@ -0,0 +1,231 @@
+  not_lt_bot
+
+@[simp]
+theorem pos_of_ne_zero (o : OrderType) : 0 < o → o ≠ 0 :=
```
This should be an iff!

YanYablonovskiy committed:
Adding `Order.Types.Defs` and moving `LinOrd` definition

YanYablonovskiy committed:
Remove `LinOrd` from old location

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>

YanYablonovskiy committed:
Update Mathlib/Order/Types/Defs.lean

Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>

YanYablonovskiy committed:
Review suggestions

YanYablonovskiy committed:
Run `mk_all`

YanYablonovskiy committed:
Add docs to instance

Even though it is local and not exposed

YanYablonovskiy committed:
Apply review suggestions

YanYablonovskiy committed:
One more suggestion

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>

YanYablonovskiy committed:
Update Defs.lean

YanYablonovskiy committed:
Remove spurious spacing

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Violeta Hernández Palacios <vi.hdz.p@gmail.com>

YanYablonovskiy committed:
Merge branch 'leanprover-community:master' into feat/OrderType_defs

YanYablonovskiy committed:
Apply suggestions from code review

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

YanYablonovskiy committed:
Remove simp tag

YanYablonovskiy committed:
Fix over 100 char line

YanYablonovskiy committed:
Add `type_nat_eq` as per suggestion

YanYablonovskiy committed:
Update Mathlib/Order/Types/Defs.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

YanYablonovskiy committed:
Adding `no_expose` to `OrderType.type`

As per suggestions, to avoid exposing `LinOrd`.

YanYablonovskiy committed:
Adding `simp` tag to `type_ne_zero`

YanYablonovskiy committed:
Update Defs.lean

YanYablonovskiy committed:
Typo

YanYablonovskiy committed:
Update Defs.lean

YanYablonovskiy committed:
Code suggestions

YanYablonovskiy committed:
Merge branch 'leanprover-community:master' into feat/OrderType_defs

YanYablonovskiy committed:
Fix `Data.Sum.Order` conflict

Try to fix whatever just happened

YanYablonovskiy committed:
Fix `Mathlib.lean`

YanYablonovskiy committed:
Partial fix

YanYablonovskiy committed:
Fix via `expose`

Ended up exposing some of the original additions to work with Lex better.

YanYablonovskiy committed:
Universe Fix attempt

YanYablonovskiy commented:
-awaiting-author

YanYablonovskiy committed:
Revert `expose` and run `min-imports`

YanYablonovskiy committed:
Remove `#min_imports` line

Yikes

YaelDillies commented on Mathlib/Order/Types/Basic.lean:
What about naming this `.Arithmetic`?

YaelDillies reviewed (approved):
Thanks! 🚀

maintainer merge

YanYablonovskiy committed:
Renaming from `Basic` to `Arithmetic`

YanYablonovskiy committed:
Adjust a few spacings

riccardobrasca commented:
This is marked as draft!

b-mehta reviewed (commented):
Nice, thank you!

bors merge
